src/hotspot/cpu/aarch64/macroAssembler_aarch64.hpp

```diff
@@ -197,8 +197,11 @@ class MacroAssembler: public Assembler {
   inline void cmp(Register Rd, unsigned char imm8)  { subs(zr, Rd, imm8); }
   inline void cmp(Register Rd, unsigned imm) = delete;
 
-  inline void cmnw(Register Rd, unsigned imm) { addsw(zr, Rd, imm); }
-  inline void cmn(Register Rd, unsigned imm) { adds(zr, Rd, imm); }
+  template<class T>
```

Comment:
Why do we need it to be a template?

Reply:
Thank you for the explanation. I reread the code and found why I got confused.
`MacroAssembler::cmnw` call `addsw` which is defined by `WRAP(addsw, true)`. 
The macro expends to
```
void addsw(Register Rd, Register Rn, uint64_t imm) {
    wrap_adds_subs_imm_insn(Rd, Rn, imm, &Assembler::addsw, &Assembler::addsw, true);
}

void addsw(Register Rd, Register Rn, Register Rm, enum shift_kind kind, unsigned shift = 0) {
   Assembler::addsw(Rd, Rn, Rm, kind, shift);
}

void addsw(Register Rd, Register Rn, Register Rm) {
   Assembler::addsw(Rd, Rn, Rm);
}

void addsw(Register Rd, Register Rn, Register Rm, ext::operation option, int amount = 0) {
  Assembler::addsw(Rd, Rn, Rm, option, amount);
}
```

I've mixed it up with `Assembler::addsw(Register Rd, Register Rn, unsigned imm)`:
```
#define INSN(NAME, decode, negated)                                     \
  void NAME(Register Rd, Register Rn, unsigned imm, unsigned shift) {   \
    starti;                                                             \
    f(decode, 31, 29), f(0b10001, 28, 24), f(shift, 23, 22), f(imm, 21, 10); \
    zrf(Rd, 0), srf(Rn, 5);                                             \
  }                                                                     \
                                                                        \
  void NAME(Register Rd, Register Rn, unsigned imm) {                   \
    starti;                                                             \
    add_sub_immediate(current_insn, Rd, Rn, imm, decode, negated);      \
  }

  INSN(addsw, 0b001, 0b011);
```